### ISSUE TITLE
fliplr_params to flip the parameters of the left hand is not giving me the right results

### ISSUE THREAD

**Issue by HaziqRazali** [geopavlakos/hamer#78]
I used HaMeR to estimate the hand pose, and it worked well, as shown in the figure on the left. I then combined the results from HaMeR and 4D-Humans into the SMPLH model, but I was unable to get the correct reconstruction for the left hand, as you can see in the figure on the right.

![hamer](https://github.com/user-attachments/assets/0b3e6f34-f974-40b2-92ce-067680c8ea07) ![image](https://github.com/user-attachments/assets/d0f799f3-f603-41c7-a33e-38544c09d437)

I followed your advice from these two threads https://github.com/geopavlakos/hamer/issues/26 https://github.com/geopavlakos/hamer/issues/61 to convert the left hand pose from a rotation matrix to axis-angle, used the [fliplr_params](https://github.com/geopavlakos/hamer/blob/ec7c9e99a24c4f2e328b7e1deda32214b06118ee/hamer/datasets/utils.py#L378C5-L378C18) function, and then converted it back to a rotation matrix. However, the results are still incorrect, as shown in the figure. I have tried flipping and no flipping but to no avail.

Do you any idea why? Would greatly appreciate your help.

**Comment by HaziqRazali** [comment]
Thank you for responding! Here are the relevant functions in use: `rotmat_to_aa`, `aa_to_rotmat`, and `fliplr_params`.

```
from scipy.spatial.transform import Rotation as R
def rotmat_to_aa(rot_matrices):
    """
    Convert a batch of rotation matrices to axis-angle format using scipy.
    
    Parameters:
    rot_matrices: numpy array of shape (batch_size, 3, 3)
                  Batch of rotation matrices
    
    Returns:
    axis_angles: numpy array of shape (batch_size, 3)
                 Batch of axis-angle vectors
    """
    # Create a Rotation object from the rotation matrices
    rotation_obj = R.from_matrix(rot_matrices)
    
    # Convert to axis-angle (rotation vector) representation
    axis_angles = rotation_obj.as_rotvec()  # This gives the axis-angle representation

    return axis_angles
    
def aa_to_rotmat(axis_angles):
    """
    Convert a batch of axis-angle vectors to rotation matrices using scipy.
    
    Parameters:
    axis_angles: numpy array of shape (batch_size, 3)
                 Batch of axis-angle vectors (rotation vector format)
    
    Returns:
    rot_matrices: numpy array of shape (batch_size, 3, 3)
                  Batch of rotation matrices
    """
    # Create a Rotation object from the axis-angle vectors
    rotation_obj = R.from_rotvec(axis_angles)
    
    # Convert to rotation matrices
    rot_matrices = rotation_obj.as_matrix()  # This gives the rotation matrices

    return rot_matrices
    
# to flip the lr params
from typing import Dict, Tuple
def fliplr_params(mano_params: Dict, has_mano_params: Dict) -> Tuple[Dict, Dict]:
    """
    Flip MANO parameters when flipping the image.
    Args:
        mano_params (Dict): MANO parameter annotations.
        has_mano_params (Dict): Whether MANO annotations are valid.
    Returns:
        Dict, Dict: Flipped MANO parameters and valid flags.
    """
    global_orient = mano_params['global_orient'].copy()
    hand_pose = mano_params['hand_pose'].copy()
    betas = mano_params['betas'].copy()
    has_global_orient = has_mano_params['global_orient'].copy()
    has_hand_pose = has_mano_params['hand_pose'].copy()
    has_betas = has_mano_params['betas'].copy()

    global_orient[1::3] *= -1
    global_orient[2::3] *= -1
    hand_pose[1::3] *= -1
    hand_pose[2::3] *= -1

    mano_params = {'global_orient': global_orient.astype(np.float32),
                   'hand_pose': hand_pose.astype(np.float32),
                   'betas': betas.astype(np.float32)
                  }

    has_mano_params = {'global_orient': has_global_orient,
                       'hand_pose': has_hand_pose,
                       'betas': has_betas
                      }

    return mano_params#, has_mano_params
```

Here's how I do the flipping: 
- I deepcopy the output of mano `mano_data` whose entries have all been converted from tensors to numpy arrays.
- Convert from rotation matrix to axis angle
- Perform the flip for only the left hand using the function you provided
- Convert it back to rotation matrix.
- Copy the data back into `mano_data`

```
temp_mano_data  = copy.deepcopy(mano_data)
lhand_mano_data = {}
lhand_mano_data["betas"]            = temp_mano_data["pred_mano_params"]["betas"][0]            # [2,  10]
lhand_mano_data["global_orient"]    = temp_mano_data["pred_mano_params"]["global_orient"][0]    # [1,  3, 3]
lhand_mano_data["hand_pose"]        = temp_mano_data["pred_mano_params"]["hand_pose"][0]        # [15, 3, 3]
lhand_mano_data["hand_pose"]        = rotmat_to_aa(lhand_mano_data["hand_pose"])                # [15, 3]
lhand_mano_data                     = fliplr_params(lhand_mano_data, lhand_mano_data)
lhand_mano_data["hand_pose"]        = aa_to_rotmat(lhand_mano_data["hand_pose"])                # [15, 3, 3]
mano_data["pred_mano_params"]["hand_pose"][0] = lhand_mano_data["hand_pose"]
```

And then I combine the updated mano params with the output of 4D-Humans to form the smplh parameters for visualization.
- I grab the output of 4D-Humans (also numpy array)
- Keep its global body orient but remove the last 2 elements of ["pred_smpl_params"]["body_pose"]

```
# from 4D-Humans
global_body_orient  = torch.tensor(smpl_data['pred_smpl_params']['global_orient'], dtype=torch.float32)   # [1, 1,  3, 3]
body_pose           = torch.tensor(smpl_data["pred_smpl_params"]["body_pose"], dtype=torch.float32)       # [1, 23, 3, 3]
body_pose           = body_pose[:,:21]                                                                              # [1, 21, 3, 3]
betas               = torch.tensor(smpl_data['pred_smpl_params']['betas'], dtype=torch.float32)           # [1, 10]

# from HaMeR
hand_pose           = torch.tensor(mano_data['pred_mano_params']['hand_pose'], dtype=torch.float32)       # [2, 15, 3, 3]

# smplh params
pred_smplh_params = {}
pred_smplh_params['betas']           = betas                # [1, 10]
pred_smplh_params['global_orient']   = global_body_orient   # [1, 1,  3, 3]
pred_smplh_params['body_pose']       = body_pose            # [1, 21, 3, 3]
pred_smplh_params['left_hand_pose']  = hand_pose[0:1]       # [1, 15, 3, 3]
pred_smplh_params['right_hand_pose'] = hand_pose[1:2]       # [1, 15, 3, 3]
```

Thank you, and I hope to seek your advice.

**Comment by geopavlakos** [comment]
`fliplr_params()` assumes that the inputs are in the form of vectors not 2D matrices (e.g., 45 dims for `hand_pose` instead of 15x3)

**Comment by HaziqRazali** [comment]
Damn that was silly of me to not notice... Thank you so much !
